pal/src/sync/CCSpinLock.cpp

```diff
@@ -0,0 +1,114 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+#include "pal/thread.hpp"
+#include "pal/cs.hpp"
+#include "pal_assert.h"
+
+#include <sched.h>
+#include <pthread.h>
+
+void CCSpinLock::Enter()
+{
+    size_t currentThreadId = -1, spin = 0;
+    if (shouldTrackThreadId)
+    {
+        currentThreadId = GetCurrentThreadId();
+        if (threadId == currentThreadId)
+        {
+            _ASSERTE(enterCount > 0);
+            enterCount++;
+            return;
+        }
+    }
+
+    while(__sync_lock_test_and_set(&lock, 1))
+    {
+        while(lock)
+        {
+            if (++spin > 11)
+            {
+                sched_yield();
+                spin = 0;
+            }
+        }
+    }
+
+    if (shouldTrackThreadId)
+    {
+        threadId = GetCurrentThreadId();
+        _ASSERTE(enterCount == 0);
+        enterCount = 1;
+    }
+}
+
+bool CCSpinLock::TryEnter()
+{
+    int spin = 0;
+    bool locked = true;
+
+    while(__sync_lock_test_and_set(&lock, 1))
```

Comment:
Why doesn't the same statement apply to "Enter"? If you can do that in "Enter", you should be able to do it in "TryEnter", no? Basically "Enter" does not spin if the thread already has the lock, but "TryEnter" still spins, why is the difference?